### DISCUSSION TITLE
Referencing relation on "grandparent" (or more distantly related) objects

### DISCUSSION THREAD

**Discussion by torstenbm** [openfga/community#16] [Q&A] [answered]
Hi!

The part of the docs, [Referencing Relations On Related Objects](https://openfga.dev/docs/configuration-language#referencing-relations-on-related-objects), do a great job of describing how to reference relations on related objects, but what if two objects are related only through a third object (or even further away)?

Imagine the following case:

We have three objects: drive, folder and doc. A user can be a member of a drive, and a drive can have a folder and a folder can have a doc.
```
type drive
  relations
    define member as self

type folder
  relations
    define parent_drive as self

type doc
  relations
    define parent_folder as self
```

Now if we want to create a relation read_folder such that only members of the parent drive of a folder have it, we would as the docs describe modify the folder type to something like this: 

```
type folder
  relations
    define parent_drive as self
    define read_folder as member from parent_drive
```

How do I go about defining a relation `read_doc` on the doc type in a way that only members of the "grandparent" drive (the drive of the folder of the doc) have it? Something along the lines of

```
type doc
  relations
    define parent_folder as self
    define read_doc as member from parent_drive from parent_folder
```
but that gives me a syntax error.

In other words: given the tuples
```
(torsten, member, drive:123)
(drive:123, parent_drive, folder:123)
(folder:123, parent_folder, doc:123)
```
To define the relation `read_folder` on the folder type such that the query `check(torsten, read_folder, folder:123)` resolves to `true` is easy.

How do I define the relation `read_doc` on the doc type such that the query `check(torsten, read_doc, doc:123)` resolves to `true`?

**Comment by torstenbm** [comment, marked as answer]
Is something like this considered best practices for these cases?

```
type drive
  relations
    define member as self

type folder
  relations
    define parent_drive as self
    define is_member_of_parent_drive as member from parent_drive
    define read_drive as is_member_of_parent_drive

type doc
  relations
    define parent_folder as self
    define is_member_of_parent_drive_of_parent_folder as is_member_of_parent_drive from parent_folder
    define read_doc as is_member_of_parent_drive_of_parent_folder
```

**Reply by rhamzeh** [reply]
Hi @torstenbm! Yes, you are correct.

The DSL does not support referencing from more distantly related objects, so the solution is exactly what you did here: bringing those relations closer.

You can see an example of this in entitlements ([Playground](https://play.fga.dev/sandbox/?store=entitlements)/[Documentation](https://openfga.dev/docs/modeling/advanced/entitlements)).

**Reply by torstenbm** [reply]
Thank you for your answer @rhamzeh! Is it just the DSL that's not supporting more distantly related objects or is it the config language itself? I.e. is it possible to reference more distantly related objects by nesting `"tupleToUserset"` attributes with JSON?
